pex/pex_bootstrapper.py

```diff
@@ -18,122 +18,159 @@
   return list(
     matched_interpreters(interpreters, compatibility_constraints)
     if compatibility_constraints
     else interpreters
   )
 
 
-def _select_pex_python_interpreter(target_python, compatibility_constraints=None):
-  target = find_in_path(target_python)
-
-  if not target:
-    die('Failed to find interpreter specified by PEX_PYTHON: %s' % target)
-  if compatibility_constraints:
-    pi = PythonInterpreter.from_binary(target)
-    if not list(matched_interpreters([pi], compatibility_constraints)):
-      die('Interpreter specified by PEX_PYTHON (%s) is not compatible with specified '
-          'interpreter constraints: %s' % (target, str(compatibility_constraints)))
-  if not os.path.exists(target):
-    die('Target interpreter specified by PEX_PYTHON %s does not exist. Exiting.' % target)
-  return target
+def _select_pex_python_interpreter(pex_python, compatibility_constraints=None):
+  compatible_interpreters = _filter_compatible_interpreters(
+    _find_pex_python(pex_python),
+    compatibility_constraints=compatibility_constraints
+  )
+  if not compatible_interpreters:
+    die('Failed to find a compatible PEX_PYTHON={} for constraints: {}'
+        .format(pex_python, compatibility_constraints))
+  return _select_interpreter(compatible_interpreters)
 
 
-def _select_interpreter(pex_python_path=None, compatibility_constraints=None):
+def _select_path_interpreter(path=None, compatibility_constraints=None):
   compatible_interpreters = find_compatible_interpreters(
-    pex_python_path=pex_python_path, compatibility_constraints=compatibility_constraints)
-
+    path=path,
+    compatibility_constraints=compatibility_constraints
+  )
   if not compatible_interpreters:
-    die('Failed to find compatible interpreter for constraints: %s'
-        % str(compatibility_constraints))
-  # TODO: https://github.com/pantsbuild/pex/issues/430
-  target = min(compatible_interpreters).binary
+    die('Failed to find compatible interpreter on path {} for constraints: {}'
+        .format(path or os.getenv('PATH'), compatibility_constraints))
+  return _select_interpreter(compatible_interpreters)
 
-  if os.path.exists(target):
-    return target
 
+def _select_interpreter(candidate_interpreters):
+  current_interpreter = PythonInterpreter.get()
+  if current_interpreter in candidate_interpreters:
+    # Always prefer continuing with the current interpreter when possible.
+    return current_interpreter
+  else:
+    # TODO: Allow the selection strategy to be parameterized:
+    #   https://github.com/pantsbuild/pex/issues/430
+    return min(candidate_interpreters)
 
-def maybe_reexec_pex(compatibility_constraints):
-  """
-  Handle environment overrides for the Python interpreter to use when executing this pex.
+
+def maybe_reexec_pex(compatibility_constraints=None):
+  """Handle environment overrides for the Python interpreter to use when executing this pex.
```

Comment:
Is this a style change or does it change any presentations from help/pydocs, etc? Just curious.

Reply:
Just a style change - I wanted to make it clear None was an acceptable input. The general implication of `def maybe_reexec_pex(compatibility_constraints)` is that `compatibility_constraints` are required.